KAFKA-17477:Topic command integration test migrate to new test infra

TaiJuWu opened the pull request:
*More detailed description of your change,
1. New interface `aliveBrokers()` for `clusterInstance`.
2. Migrate topic command integration test migrate to new test infra

*Summary of testing strategy (including rationale)


### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

chia7712 commented on core/src/test/java/kafka/testkit/KafkaClusterTestKit.java:
```diff
@@ -188,12 +188,13 @@ private KafkaConfig createNodeConfig(TestKitNode node) {
             }
             props.put(SocketServerConfigs.LISTENER_SECURITY_PROTOCOL_MAP_CONFIG,
-                    "EXTERNAL:PLAINTEXT,CONTROLLER:PLAINTEXT");
+                    "EXTERNAL:PLAINTEXT,CONTROLLER:PLAINTEXT,PLAINTEXT:PLAINTEXT");
```
why we need this change?

TaiJuWu replied:
The TestUtils requires sever use `PLAINTEXT` protocol, you can reference [here.](https://github.com/apache/kafka/blob/d6cd83e2fb2bab4526f07e067277b34e482f6678/core/src/test/scala/unit/kafka/utils/TestUtils.scala#L232)

chia7712 replied:
not sure why we need to call that test helper?

TaiJuWu replied:
In this PR, I try to minimal this patch so I don't touch any helper function.

chia7712 replied:
that is good. If you feel this PR needs some helpers, we can have another PR to address that.

chia7712 commented on core/src/test/java/kafka/testkit/KafkaClusterTestKit.java:
```diff
@@ -353,6 +354,9 @@ static private void setupNodeDirectories(File baseDirectory,
     private final Map<Integer, ControllerServer> controllers;
     private final Map<Integer, BrokerServer> brokers;
+
+    private final Map<Integer, Boolean> aliveBrokers;
```
Instead of tracing alive brokers, maybe we should add a new method `isShutdown` to `KafkaBroker`?

TaiJuWu replied:
Sounds great but if a broker already shutdown, how can we access the server and check its status?

chia7712 replied:
> if a broker already shutdown, how can we access the server and check its status?

both zk and kraft broker has in-memory state, so maybe you can leverage them?

chia7712 reviewed (commented):
@TaiJuWu thanks for this patch

chia7712 commented:
@TaiJuWu please rebase code

TaiJuWu commented:
> @TaiJuWu please rebase code

Rebase already, thanks.

TaiJuWu committed:
Add getCurrentFunctionName function

TaiJuWu committed:
Rewrite Topic command itergration test to new infra

TaiJuWu committed:
Fix test fail

Old testcase use PLAINTEXT as its protocol, I try to minimize this PR
so make TestKit support PLAINTEXT protocol.

TaiJuWu committed:
Wait all brokers to start running

TaiJuWu committed:
fix build error

chia7712 commented on core/src/test/java/kafka/testkit/KafkaClusterTestKit.java:
```diff
@@ -104,14 +104,14 @@ synchronized void registerPort(int nodeId, int port) {
-                                Map.Entry::getKey,
-                                entry -> new InetSocketAddress("localhost", entry.getValue())
-                            )
+                                Collectors.toMap(
```
@TaiJuWu please revert those changes :)

chia7712 commented on tools/src/test/java/org/apache/kafka/tools/TopicCommandIntegrationTest.java:
```diff
@@ -86,652 +85,922 @@
+        String testTopicName = TestUtils.getCurrentFunctionName() + "-" + TestUtils.randomString(10);
+
+        try (Admin adminClient = clusterInstance.createAdminClient()) {
+            Buffer<KafkaBroker> scalaBrokers = JavaConverters.asScalaIteratorConverter(
```
please avoid using scala collections

chia7712 commented on tools/src/test/java/org/apache/kafka/tools/TopicCommandIntegrationTest.java:
```diff
@@ -86,652 +85,952 @@
-                "Config not set correctly: " + configs.get("delete.retention.ms").value());
+    @ClusterTemplate("generate1")
+    public void testCreateWithConfigs() throws Exception {
+        String testTopicName = org.apache.kafka.test.TestUtils.getCurrentFunctionName() + "-" +
```
It seems to me `randomString` is good enough. Or we can add a static count to generate random topic name

TaiJuWu replied:
My consideration is the FunctionName can help us to address which testcase fail.
WDYT?

chia7712 replied:
or we can inject `TestInfo` to get name?

chia7712 commented on core/src/test/java/kafka/testkit/KafkaClusterTestKit.java:
```diff
@@ -356,18 +357,19 @@ private static void setupNodeDirectories(File baseDirectory,
-        ControllerQuorumVotersFutureManager controllerQuorumVotersFutureManager,
-        File baseDirectory,
-        SimpleFaultHandlerFactory faultHandlerFactory
+            ExecutorService executorService,
```
please revert those changes

TaiJuWu replied:
Ok. Please wait a moment.

chia7712 commented on core/src/test/java/kafka/testkit/KafkaClusterTestKit.java:
```diff
@@ -390,7 +392,7 @@ public void format() throws Exception {
                 futures.add(executorService.submit(() -> {
                     formatNode(broker.sharedServer().metaPropsEnsemble(),
-                        !nodes().brokerNodes().get(entry.getKey()).combined());
+                            !nodes().brokerNodes().get(entry.getKey()).combined());
```
ditto

chia7712 commented on core/src/test/java/kafka/test/junit/RaftClusterInvocationContext.java:
```diff
@@ -189,7 +189,9 @@ public void start() {
                 if (started.compareAndSet(false, true)) {
                     clusterTestKit.startup();
                     kafka.utils.TestUtils.waitUntilTrue(
-                            () -> this.clusterTestKit.brokers().get(0).brokerState() == BrokerState.RUNNING,
```
nice finding. Maybe you can file a MINOR patch for it, since this PR needs more time to be completed

TaiJuWu replied:
I open a new PR for this change.
Please take a look [here](https://github.com/apache/kafka/pull/16537)

chia7712 reviewed (commented):
@TaiJuWu thanks for your patch

TaiJuWu committed:
remove scala collection dependency

TaiJuWu commented:
There are some tests fail, I am going to dig it.

TaiJuWu committed:
Revert "Fix test fail"

This reverts commit 779501a8720e9a5e31fc2859347c2bdb04d495b4.

TaiJuWu committed:
KafkaClusterTestKit support PLAINTEXT
Old testframe use PLAINTEXT as protocol so we need to support it for compability.

TaiJuWu committed:
Revert "Wait all brokers to start running"

This reverts commit 884e1d82dd85b17becbe065ba140e0b011b7c388.

TaiJuWu committed:
Fix testDeleteInternalTopic

TaiJuWu committed:
Fix Zk mode testTopicWithCollidingCharDeletionAndCreateAgain

TaiJuWu committed:
Fix testDescribeUnderReplicatedPartitionsWhenReassignmentIsInProgress

TaiJuWu committed:
fix build fail

chia7712 commented on clients/src/test/java/org/apache/kafka/test/TestUtils.java:
```diff
@@ -706,4 +706,9 @@ public static ApiVersionsResponse createApiVersionsResponse(
                 zkMigrationEnabled);
     }
+
+    public static String getCurrentFunctionName() {
```
IIRC, we had a discussion about the topic name. What is the updates?

TaiJuWu replied:
Sorry, I missed the newset commend.
Will do ASAP.

TaiJuWu replied:
Unfortunately ,we can't use `testInfo`, use `testCreateWithDefaults` as example, the name from `testInfo.getDisplayName()` will get `testCreateWithDefaults [1] Type=Raft-Isolated, MetadataVersion=4.0-IV0,Security=PLAINTEXT` and it is illegal name.

chia7712 replied:
Have you toke a look at `testInfo.getTestMethod()`?

TaiJuWu replied:
> Have you toke a look at `testInfo.getTestMethod()`?

Yes, they are same problem.

chia7712 replied:
> Yes, they are same problem.

Have you test it on local? `testInfo.getTestMethod().get().getName()` show the "pure" method name on my local.

TaiJuWu replied:
Thanks for your info. Update already.

chia7712 commented on core/src/test/java/kafka/testkit/KafkaClusterTestKit.java:
```diff
@@ -191,7 +191,7 @@ private KafkaConfig createNodeConfig(TestKitNode node) {
             }
             props.put(SocketServerConfigs.LISTENER_SECURITY_PROTOCOL_MAP_CONFIG,
-                    "EXTERNAL:PLAINTEXT,CONTROLLER:PLAINTEXT");
+                    "EXTERNAL:PLAINTEXT,CONTROLLER:PLAINTEXT,PLAINTEXT:PLAINTEXT");
```
Do we still need those changes?

TaiJuWu replied:
I will check again and try to remove .
Thanks for your review.

TaiJuWu replied:
revert, thanks for your remind.

chia7712 reviewed (commented):
@TaiJuWu thanks for updates

TaiJuWu committed:
Revert "KafkaClusterTestKit support PLAINTEXT"

This reverts commit bfd38cf81f2803a019075f7309d276dc8812f772.

TaiJuWu committed:
Revert "Add getCurrentFunctionName function" and use testInfo

This reverts commit d782cb843fb9f584c2c5c14a37c5b29548de43d5.

chia7712 commented on tools/src/test/java/org/apache/kafka/tools/TopicCommandIntegrationTest.java:
```diff
@@ -86,652 +90,792 @@
-            String deletePath = kafka.zk.DeleteTopicsTopicZNode.path(topicWithCollidingChar);
-            assertFalse(zkClient().pathExists(deletePath), "Delete path for topic shouldn't exist before deletion.");
+    @ClusterTemplate("generate1")
+    public void testTopicWithCollidingCharDeletionAndCreateAgain(TestInfo testInfo) throws Exception {
```
Please remove `testInfo` if the test does not use it.

chia7712 commented on tools/src/test/java/org/apache/kafka/tools/TopicCommandIntegrationTest.java:
```diff
@@ -86,652 +90,792 @@
+                    60000, "Delete topic fail in 60000 ms"
+            );
+
+            kafka.utils.TestUtils.verifyTopicDeletion(null, topicWithCollidingChar, 1, JavaConverters.asScalaIteratorConverter(
```
Could you please file a MINOR patch to add `verifyTopicDeletion` to `ClusterInstance`?

TaiJuWu replied:
Ok. I am working on this.

TaiJuWu replied:
I opened a PR [here]( https://github.com/apache/kafka/pull/16627/commits/3883c4b2115be2c59c86e51fc8b67648321aa4df). Please take a look.

chia7712 commented on tools/src/test/java/org/apache/kafka/tools/TopicCommandIntegrationTest.java:
```diff
@@ -86,652 +90,792 @@
+            clusterInstance.shutdownBroker(0);
+            assertEquals(5, clusterInstance.aliveBrokers().size());
 
             // wait until the topic metadata for the test topic is propagated to each alive broker
```
It seems this is equal to `ClusterInstance#waitForTopic`, since it checks the metadata cache for the specific topic/partition

TaiJuWu replied:
Agree, update. Thanks.

chia7712 reviewed (commented):
@TaiJuWu thanks for updated patch!

TaiJuWu committed:
Merge branch 'trunk' into TopicComIT-migrate

TaiJuWu committed:
remove unnecessary testinfo

TaiJuWu committed:
fix testDescribeUnavailablePartitions

TaiJuWu commented on core/src/test/java/kafka/test/ClusterInstance.java:
```diff
@@ -188,15 +188,15 @@ default Set<GroupProtocol> supportedGroupProtocols() {
         // wait for metadata
         TestUtils.waitForCondition(
-            () -> brokers().values().stream().allMatch(broker -> partitions == 0 ?
+            () -> aliveBrokers().values().stream().allMatch(broker -> partitions == 0 ?
```
If we don't have this modification, isolated-raft always failed.
Maybe it is waiting the shutdown broker?

chia7712 commented:
@TaiJuWu Could you please rebase code to trigger QA again?

TaiJuWu committed:
Merge branch 'trunk' into TopicComIT-migrate

TaiJuWu committed:
apply cluster.waitTopicDeletion

TaiJuWu commented:
> @TaiJuWu Could you please rebase code to trigger QA again?

Done. Thanks!

chia7712 commented on tools/src/test/java/org/apache/kafka/tools/TopicCommandIntegrationTest.java:
```diff
@@ -85,653 +80,761 @@
 import static org.mockito.Mockito.spy;
 
 @Tag("integration")
```
Could you please remove this tag? it is redundant to `ClusterTestExtensions`, as it is added automatically.

chia7712 commented on tools/src/test/java/org/apache/kafka/tools/TopicCommandIntegrationTest.java:
```diff
@@ -85,653 +80,761 @@
-@SuppressWarnings("deprecation") // Added for Scala 2.12 compatibility for usages of JavaConverters
-public class TopicCommandIntegrationTest extends kafka.integration.KafkaServerTestHarness implements Logging, RackAwareTest {
+@ExtendWith(ClusterTestExtensions.class)
+public class TopicCommandIntegrationTest {
```
@TaiJuWu Could you please merge `TopicCommandTest` and `TopicCommandIntegrationTest` into single one? `ClusterTest` and `Test` can be existent in the same class

TaiJuWu replied:
Will do this weekend, Thanks!

TaiJuWu replied:
Updated already, please take a look, thanks!

chia7712 commented on tools/src/test/java/org/apache/kafka/tools/TopicCommandIntegrationTest.java:
```diff
@@ -85,653 +80,761 @@
-                "Admin client didn't see the created topic. It saw: " + adminClient.listTopics().names().get());
+    @ClusterTemplate("generate1")
+    public void testCreate(TestInfo testInfo) throws InterruptedException, ExecutionException {
+        String testTopicName = testInfo.getTestMethod().get().getName() + "-" +
```
not sure why we need the postfix? each test case has its own embedded cluster, so there is no name conflicting I feel :_

chia7712 commented on tools/src/test/java/org/apache/kafka/tools/TopicCommandIntegrationTest.java:
```diff
@@ -85,653 +80,761 @@
-            .partitions();
-        assertEquals(defaultNumPartitions, partitions.size(), "Unequal partition size: " + partitions.size());
-        assertEquals(defaultReplicationFactor, (short) partitions.get(0).replicas().size(), "Unequal replication factor: " + partitions.get(0).replicas().size());
+    @ClusterTemplate("generate1")
```
`generate1` -> `generate`

chia7712 commented on tools/src/test/java/org/apache/kafka/tools/TopicCommandIntegrationTest.java:
```diff
@@ -85,653 +80,761 @@
+        rackInfo.put(5, infoPerBroker6);
+
+        return Collections.singletonList(ClusterConfig.defaultBuilder()
+                .setBrokers(6)
```
Maybe we can test "zk and "kraft" instead of all types.

chia7712 commented on tools/src/test/java/org/apache/kafka/tools/TopicCommandIntegrationTest.java:
```diff
@@ -1161,4 +1278,32 @@ public ReplicaDistributions(Map<Integer, List<String>> partitionRacks,
+        return new KafkaProducer<>(producerProps, new StringSerializer(), new StringSerializer());
+    }
+
+    private void sendProducerRecords(String testTopicName, KafkaProducer<String, String> producer, Integer numMessage) {
```
`Integer` -> int

chia7712 commented on tools/src/test/java/org/apache/kafka/tools/TopicCommandIntegrationTest.java:
```diff
@@ -1161,4 +1278,32 @@ public ReplicaDistributions(Map<Integer, List<String>> partitionRacks,
+        List<ProducerRecord<String, String>> records = generateProduceMessage(testTopicName, numMessage);
+        List<Future<RecordMetadata>> features = records.stream().map(producer::send).collect(Collectors.toList());
+
+        assertDoesNotThrow(() -> features.forEach(s -> {
```
```java
    private void sendProducerRecords(String testTopicName, KafkaProducer<String, String> producer, Integer numMessage) {
        IntStream.range(0, numMessage).forEach(i -> producer.send(new ProducerRecord<>(testTopicName, "test-" + i)));
        producer.flush();
    }
```

chia7712 commented on tools/src/test/java/org/apache/kafka/tools/TopicCommandIntegrationTest.java:
```diff
@@ -740,175 +843,181 @@ public void testDescribeUnavailablePartitions(String quorum) throws ExecutionExc
+                        try {
+                            return !adminClient.listPartitionReassignments(Collections.singleton(tp)).reassignments().get()
+                                    .get(tp).addingReplicas().isEmpty();
+                        } catch (InterruptedException | ExecutionException e) {
```
we don't need this try-catch, right?

OmniaGM replied:
this can be 
```
TestUtils.waitForCondition(
                    () -> !adminClient.listPartitionReassignments(Collections.singleton(tp)).reassignments().get()
                        .get(tp).addingReplicas().isEmpty(), 60000, "Reassignment didn't add the second node"
            );
 ```

no need to wrap it in try and catch.

chia7712 commented on tools/src/test/java/org/apache/kafka/tools/TopicCommandIntegrationTest.java:
```diff
@@ -740,175 +843,181 @@ public void testDescribeUnavailablePartitions(String quorum) throws ExecutionExc
+                        try {
+                            PartitionReassignment tempReassignments = adminClient.listPartitionReassignments(Collections.singleton(tp)).reassignments().get().get(tp);
+                            reassignmentsRef.set(tempReassignments);
+                        } catch (InterruptedException | ExecutionException e) {
```
we don't need try-catch, right?

chia7712 commented on tools/src/test/java/org/apache/kafka/tools/TopicCommandIntegrationTest.java:
```diff
@@ -85,653 +80,761 @@
             TopicDescription testTopicDescription = adminClient.describeTopics(Collections.singletonList(testTopicName))
-                .allTopicNames().get().get(testTopicName);
+                    .allTopicNames().get().get(testTopicName);
             int partitionOnBroker0 = testTopicDescription.partitions().stream()
```
this variable is unused. could you please remove it

chia7712 reviewed (commented):
@TaiJuWu thanks for this patch

OmniaGM commented on tools/src/test/java/org/apache/kafka/tools/TopicCommandIntegrationTest.java:
```diff
@@ -85,653 +80,761 @@
-        testTopicName = String.format("%s-%s", info.getTestMethod().get().getName(), org.apache.kafka.test.TestUtils.randomString(10));
-        scalaBrokers = brokers();
-        scalaControllers = controllerServers();
+    static List<ClusterConfig> generate1() {
```
why it is called `generate1` if there is no any others?

TaiJuWu replied:
There is no reason, we also can call other name!

OmniaGM commented on tools/src/test/java/org/apache/kafka/tools/TopicCommandIntegrationTest.java:
```diff
@@ -85,653 +80,761 @@
-        String output = captureListTopicStandardOut(buildTopicCommandOptionsWithBootstrap("--list", "--topic", "kafka.*"));
+    @ClusterTemplate("generate1")
+    public void testListTopicsWithExcludeInternal() throws InterruptedException {
+        try (Admin adminClient = clusterInstance.createAdminClient();) {
```
There is unnecessary semicolon here

OmniaGM commented on tools/src/test/java/org/apache/kafka/tools/TopicCommandIntegrationTest.java:
```diff
@@ -85,653 +80,761 @@
+            String topic3 = "oooof.testTopic1";
+            int partition = 2;
+            short replicationFactor = 2;
+            adminClient.createTopics(Collections.singletonList(new NewTopic(topic1, partition, replicationFactor)));
```
Can we place all these with 
```
  for (String topic: Arrays.asList(topic1, topic2, topic3)) {
                adminClient.createTopics(Collections.singletonList(new NewTopic(topic, partition, replicationFactor)));
                clusterInstance.waitForTopic(topic, partition);
            }
```

OmniaGM replied:
Also it look like we create topic and wait for it maybe we should have a helper that create topic and wait either in this test or in `clusterInstance`

TaiJuWu replied:
Got it, will fill another PR.

TaiJuWu replied:
File another PR, https://github.com/apache/kafka/pull/16852, PTAL.

OmniaGM commented on tools/src/test/java/org/apache/kafka/tools/TopicCommandIntegrationTest.java:
```diff
@@ -85,653 +80,761 @@
+        String testTopicName = testInfo.getTestMethod().get().getName() + "-" +
+                TestUtils.randomString(10);
+        try (Admin adminClient = clusterInstance.createAdminClient();
+             TopicCommand.TopicService topicService = new TopicCommand.TopicService(adminClient);) {
```
Unnecessary semicolon

OmniaGM commented on tools/src/test/java/org/apache/kafka/tools/TopicCommandIntegrationTest.java:
```diff
@@ -85,653 +80,761 @@
+            topicService.deleteTopic(deleteOffsetTopicOpts);
+            TestUtils.waitForCondition(
+                    () -> adminClient.listTopics().listings().get().stream().noneMatch(topic -> topic.name().equals(Topic.GROUP_METADATA_TOPIC_NAME)),
+                    60000, "Delete topic fail in 60000 ms"
```
I can see that `60000` is set as max time out in many places can we extract this into a contestant instead?

OmniaGM commented on tools/src/test/java/org/apache/kafka/tools/TopicCommandIntegrationTest.java:
```diff
@@ -740,175 +843,181 @@ public void testDescribeUnavailablePartitions(String quorum) throws ExecutionExc
+
+            TestUtils.waitForCondition(
+                    () -> {
+                        try {
```
ditto

OmniaGM commented on tools/src/test/java/org/apache/kafka/tools/TopicCommandIntegrationTest.java:
```diff
@@ -921,163 +1030,169 @@ public void testDescribeAtMinIsrPartitions(String quorum) {
+            newTopics.add(new NewTopic(fullyReplicatedTopic, fullyReplicatedReplicaAssignmentMap));
+
+            adminClient.createTopics(newTopics);
+            clusterInstance.waitForTopic(underMinIsrTopic, partitions);
```
can we replace all these `waitForTopic` into 
```
for (NewTopic topioc: newTopics) {
                clusterInstance.waitForTopic(topioc.name(), partitions);
            }
    ```

OmniaGM commented on tools/src/test/java/org/apache/kafka/tools/TopicCommandIntegrationTest.java:
```diff
@@ -921,163 +1030,169 @@ public void testDescribeAtMinIsrPartitions(String quorum) {
+            Assertions.assertEquals(5, clusterInstance.aliveBrokers().size());
+
+            TestUtils.waitForCondition(
+                    () -> clusterInstance.aliveBrokers().values().stream().allMatch(broker -> broker.metadataCache().getPartitionInfo(underMinIsrTopic, 0).get().isr().size() < 6 &&
```
this line is too long can we break this into multiple lines. Maybe something like this 
```
TestUtils.waitForCondition(
                    () -> clusterInstance.aliveBrokers().values().stream()
                            .allMatch(broker ->
                                    broker.metadataCache().getPartitionInfo(underMinIsrTopic, 0).get().isr().size() < 6 &&
                                    broker.metadataCache().getPartitionInfo(offlineTopic, 0).get().leader() == MetadataResponse.NO_LEADER_ID
                            ),
                    60000, "Timeout waiting for partition metadata propagating to brokers for underMinIsrTopic topic"
            );
     ```

OmniaGM commented on tools/src/test/java/org/apache/kafka/tools/TopicCommandIntegrationTest.java:
```diff
@@ -921,163 +1030,169 @@ public void testDescribeAtMinIsrPartitions(String quorum) {
+    public void testDescribeAndListTopicsWithoutInternalTopics(TestInfo testInfo) throws InterruptedException {
+        String testTopicName = testInfo.getTestMethod().get().getName() + "-" +
+                TestUtils.randomString(10);
+        try (Admin adminClient = clusterInstance.createAdminClient();) {
```
Unnecessary semicolon

OmniaGM commented on tools/src/test/java/org/apache/kafka/tools/TopicCommandIntegrationTest.java:
```diff
@@ -1111,7 +1227,8 @@ private String captureDescribeTopicStandardOut(TopicCommand.TopicCommandOptions
         Runnable runnable = () -> {
-            try {
+            try (Admin adminClient = clusterInstance.createAdminClient();
+                 TopicCommand.TopicService topicService = new TopicCommand.TopicService(adminClient);) {
```
Unnecessary semicolon

OmniaGM reviewed (changes requested):
Thanks for the PR, I had a quick Look and left few comments

TaiJuWu committed:
address comments

TaiJuWu committed:
clusterInstance comes from parameter

TaiJuWu committed:
move TopicCommandTest to TopicCommandIntegrationTest

chia7712 commented:
@TaiJuWu Could you please fix the conflicts?

TaiJuWu commented:
> @TaiJuWu Could you please fix the conflicts?

Hi @chia7712 , thanks for review, update!

TaiJuWu committed:
Merge branch 'trunk' into TopicComIT-migrate

chia7712 commented on tools/src/test/java/org/apache/kafka/tools/TopicCommandIntegrationTest.java:
```diff
@@ -64,851 +74,1143 @@
-@SuppressWarnings("deprecation") // Added for Scala 2.12 compatibility for usages of JavaConverters
-public class TopicCommandIntegrationTest extends kafka.integration.KafkaServerTestHarness implements Logging, RackAwareTest {
+@ExtendWith(ClusterTestExtensions.class)
+public class TopicCommandIntegrationTest {
```
@TaiJuWu Could you please rename it to `TopicCommandTest`?

TaiJuWu replied:
Done.

TaiJuWu committed:
Rename

chia7712 commented on tools/src/test/java/org/apache/kafka/tools/TopicCommandTest.java:
```diff
@@ -293,4 +332,1151 @@ public void assertInitializeInvalidOptionsExitCode(int expected, String[] option
+        Map<Integer, Integer> partitionCount = new HashMap<>();
+        Map<Integer, List<String>>  partitionRackMap = new HashMap<>();
+
+        assignment.entrySet().stream().forEach(entry -> {
```
```
        assignment.forEach((partitionId, replicaList) -> {
            Integer leader = replicaList.get(0);
            leaderCount.put(leader, leaderCount.getOrDefault(leader, 0) + 1);
            replicaList.forEach(brokerId -> {
                partitionCount.put(brokerId, partitionCount.getOrDefault(brokerId, 0) + 1);
                String rack;
                if (brokerRackMapping.containsKey(brokerId)) {
                    rack = brokerRackMapping.get(brokerId);
                    List<String> partitionRackValues = Stream.of(Collections.singletonList(rack), partitionRackMap.getOrDefault(partitionId, Collections.emptyList()))
                            .flatMap(List::stream)
                            .collect(Collectors.toList());
                    partitionRackMap.put(partitionId, partitionRackValues);
                } else {
                    System.err.printf("No mapping found for %s in `brokerRackMapping`%n", brokerId);
                }
            });
        });
```

chia7712 commented on tools/src/test/java/org/apache/kafka/tools/TopicCommandTest.java:
```diff
@@ -293,4 +332,1151 @@ public void assertInitializeInvalidOptionsExitCode(int expected, String[] option
+                                          Boolean verifyLeaderDistribution,
+                                          Boolean verifyReplicasDistribution) {
+        // always verify that no broker will be assigned for more than one replica
+        assignment.entrySet().stream()
```
```
        assignment.forEach((key, value1) -> assertEquals(new HashSet<>(value1).size(), value1.size(),
                "More than one replica is assigned to same broker for the same partition"));
```

chia7712 reviewed (commented):
@TaiJuWu thanks for your patch
